fuel-vm/src/interpreter/memory.rs

```diff
@@ -19,23 +22,324 @@ use fuel_types::{
     Word,
 };
 
-use alloc::boxed::Box;
-use core::{
-    ops,
-    ops::Range,
+use core::ops::Range;
+
+#[cfg(any(test, feature = "test-helpers"))]
+use core::ops::{
+    Index,
+    IndexMut,
+    RangeFrom,
+    RangeTo,
 };
 
-pub type Memory<const SIZE: usize> = Box<[u8; SIZE]>;
+use alloc::{
+    vec,
+    vec::Vec,
+};
 
 #[cfg(test)]
 mod tests;
 
+#[cfg(test)]
+mod impl_tests;
+
 #[cfg(test)]
 mod allocation_tests;
 
 #[cfg(test)]
 mod stack_tests;
 
+/// Resize the heap to at least `new_len` bytes, filling the new space with zeros.
+/// If `new_len` is less than the current length, the function does nothing.
+/// The function may grow the size more than `new_len` to avoid frequent
+/// reallocations.
+fn reverse_resize_at_least(vec: &mut Vec<u8>, new_len: usize) {
+    if vec.len() >= new_len {
+        return
+    }
+
+    // To reduce small allocations, allocate at least 256 bytes at once.
+    // After that, double the allocation every time.
+    let cap = new_len.next_power_of_two().clamp(256, MEM_SIZE);
+    let mut new_vec = Vec::new();
+    new_vec.reserve_exact(cap);
+    new_vec.extend(core::iter::repeat(0).take(cap - vec.len()));
+    new_vec.extend(vec.iter().copied());
+    *vec = new_vec;
+}
+
+/// The memory of the VM, represented as stack and heap.
+#[derive(Debug, Clone, PartialEq, Eq, Derivative)]
+pub struct Memory {
+    /// Stack. Grows upwards.
+    #[derivative(Debug(format_with = "fmt_truncated_hex::<16>"))]
+    stack: Vec<u8>,
+    /// Heap. Grows downwards from MEM_SIZE.
+    #[derivative(Debug(format_with = "fmt_truncated_hex::<16>"))]
+    heap: Vec<u8>,
+    /// Lowest allowed heap address, i.e. hp register value.
+    /// This is needed since we can allocate extra heap for performance reasons.
+    hp: usize,
+}
+
+impl Default for Memory {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Memory {
+    /// Create a new VM memory.
+    pub fn new() -> Self {
+        Self {
+            stack: Vec::new(),
+            heap: Vec::new(),
+            hp: MEM_SIZE,
+        }
+    }
+
+    /// Offset of the heap section
+    fn heap_offset(&self) -> usize {
+        MEM_SIZE.saturating_sub(self.heap.len())
+    }
+
+    /// Returns a linear memory representation where stack is at the beginning and heap is
+    /// at the end.
+    pub fn into_linear_memory(self) -> Vec<u8> {
+        let uninit_memory_size = MEM_SIZE
+            .saturating_sub(self.stack.len())
+            .saturating_sub(self.heap.len());
+        let uninit_memory = vec![0u8; uninit_memory_size];
+        let mut memory = self.stack;
+        memory.extend(uninit_memory);
+        memory.extend(self.heap);
+        memory
+    }
+
+    /// Grows the stack to be at least `new_sp` bytes.
+    pub fn grow_stack(&mut self, hp: Reg<HP>, new_sp: Word) -> Result<(), PanicReason> {
```

Comment:
Hmm, it seems we don't need to pass `Reg<HP>` since we have `self.hp` that represents the same. I know that you are using it in the tests to manipulate the heap, but maybe tests shouldn't affect the method API.

Reply:
True. I find it quite distasteful to even duplicate `$hp` inside this struct, but now that we do it, why not use it for this as well. Will fix.

Reply:
Done in 3baee80c